SFB-71: Fix codelijsten

JonasVanHoof opened the pull request:
## ID
[ SFB-71](https://binnenland.atlassian.net/browse/SFB-71)

 ## Description

When selecting the input type `dropdown` it didn't show the options for that `concept-scheme`. This was because we do not have all the options in our metaGraph in the local store. Thisdata has to be get in our Virtuoso store. I created a component that will query the database.

As the styling of the `codelijsten` page was not great (I alays copied `code:`)  with it what results in wrong config for the component. Updated this with use of `AuPanel`.


The options will show in the preview but not sure what has to be done when the user has selected that an option. This component is also only in the `form-builder` so if people want to use this component outside of form builder they will NOT be able to for now.

 ## Type of change

 - [x] Bug fix
 - [ ] New feature
 - [ ] Breaking change
 - [ ] Other

 ## How to test

Go to `codelijsten` endpoint select an item and see if the options display on the right. The code configuration for that `concept-scheme` should show in the panel below. Copy the code and past it in the `configuration textArea`  and than select field type to be `dropdown`.

 ## Links to other PR's

 -

aatauil commented:
Error in console:
![image](https://github.com/lblod/frontend-form-builder/assets/52280338/37f5cde3-cc09-4258-ae65-0c200d6e3666)

aatauil commented on app/routes/codelijsten.js:
```diff
@@ -1,3 +1,9 @@
+  setupController(controller, model) {
+    super.setupController(...arguments);
+
+    controller.setup(model);
```
avoid using the setupController to act as an init lifecyclehook, better place to do this is the constructor of the controller itself. also specifcally in this case, the setup function does not expect any arguments

JonasVanHoof replied:
The constructor is only called initially what means when leaving the `codelijsten` route and than coming back it is not called unfortunately. Also the [glimmer component](https://guides.emberjs.com/release/upgrading/current-edition/glimmer-components/#toc_lifecycle-and-properties) only has two lifecycle hooks to use.. Fixed this by calling the `@action updateSelectedSchemeTo` in the template so it is called everytime the template loads.

JonasVanHoof commented:
> Error in console: ![image](https://user-images.githubusercontent.com/52280338/282869510-37f5cde3-cc09-4258-ae65-0c200d6e3666.png)

Im aware of this.. there we have this issue for [SFB-45](https://binnenland.atlassian.net/browse/SFB-45) on Jira but as the functionality is now it works. The bigger problem is that we have to handle the errors coming from the RDF form/helper package/ store when a change is applied and it is invalid. (Atleast I think it is why the error is popping-up).

Will investigate :ok_hand:

JonasVanHoof commented:
@aatauil the error is not visible anymore and im checking on json format in the component now. Not showing a toaster message is actually not a possibility as this would be more handy for the user than a `console.log` message.

The problem possibly has something to do with the ember array that is being used in the `ToasterService`  [here](https://github.com/appuniversum/ember-appuniversum/blob/master/addon/services/toaster.js#L7C27-L7C27). I updated that line in the ember-appuniversum project to just `@tracked toasts = A();` instead of `@tracked toasts = A([]);` this didn't break anything.

Tried using the appuniversum local package in the form-builder application but this still returns issues with compatibility with version V1 and V2. So after trying i that and not getting it to work I changed it here to just a console.log message.

This is something that has to change but im not sure where the problem actually is...

aatauil commented on app/components/store-concept-scheme-selector.js:
```diff
@@ -0,0 +1,65 @@
+  @action
+  setSelected(value) {
+    this.selected = value;
+    // What should happen here?
```
Nothing, since this will be shown in the previewform :)

aatauil commented on app/routes/formbuilder/edit.js:
```diff
@@ -70,6 +71,11 @@ export default class FormbuilderEditRoute extends Route {
+      {
+        displayType:
+          'http://lblod.data.gift/display-types/storeConceptSchemeSelector',
+        edit: StoreConceptSchemeSelectorComponent,
```
The use of a custom component might be a bit troublesome. When the user chooses a dropdown, it will reflect in the outputed code. 

![image](https://github.com/lblod/frontend-form-builder/assets/52280338/d9f2430f-18bb-43f4-be9e-bac44b83b4ee)

I tried to overwrite the component but unfortunatly ember-submission-form-fields does not allow it so we might be a bit stuck here.

JonasVanHoof replied:
If we..

The code in the editor will always show the custom component `StoreConceptSchemeSelector`. If we add somewhere near the editor a button to export the code we than can overwrite every custom component if we want but not sure if this is something we want to do

aatauil reviewed (changes requested):
While this is a good solution, it will most likely come back to haunt us later unless we fix the root cause. This will probably need a change in ember-submission-form-fields instead of in the formbuilder

JonasVanHoof commented:
> While this is a good solution, it will most likely come back to haunt us later unless we fix the root cause. This will probably need a change in ember-submission-form-fields instead of in the formbuilder

@aatauil 
I created this [PR in ESFF](https://github.com/lblod/ember-submission-form-fields/pull/156) to validate the options passed to the `concept-scheme-selector` .

This is only step one of the solution.. are we able to call the database store from the ESFF? Else we could add a boolean `ext:fetchLocal false` to our selector turtle config what will set the way of fetching information in the component. Not sure if this is possible.

JonasVanHoof commented on app/utils/query-sparql-query.js:
```diff
@@ -0,0 +1,16 @@
+export async function queryDB(query, sparqlEndpoint = '/sparql') {
```
I created this method as it should return the json response of the query. I also saw that [here](https://github.com/lblod/frontend-form-builder/blob/master/app/services/database.js) that there is a database service that does almost the same with the difference that that method returns the full response and not the bindings as json. I could add ... this after my query but that would make it very ugly and not easy to understand.
```
let jsonResponds = await response.json();
this.options =  jsonResponds.results.bindings;
```

Just adding this comment here as we should be aware of this and maybe we can find a solution :)

JonasVanHoof commented on app/components/store-concept-scheme-selector.js:
```diff
@@ -0,0 +1,50 @@
+import { action } from '@ember/object';
+import { queryDB } from '../utils/query-sparql-query';
+
+export default class StoreConceptSchemeSelectorComponent extends Component {
```
WE should be creating the custom components for the radio & checkbox selectors to if we implement the overwrite method for the preview.

JonasVanHoof commented:
Will be removing the overwrite method and the custom component to show the options in the selectors. First will be implementing the backend and updating the frontend to work with the store instead of direct sparql queries. More info can be found here https://github.com/lblod/frontend-form-builder/pull/66

JonasVanHoof commented:
Closing this PR because of implemetation for the concept-schemes by store. https://github.com/lblod/frontend-form-builder/pull/66
